src/connectionInfo/connectionManager.ts

```diff
@@ -0,0 +1,159 @@
+import { isIPv4 } from 'net';
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+import { DEFAULT_HOST } from '../utils/constants';
+import { Disposable } from '../utils/dispose';
+import { SETTINGS_SECTION_ID, SettingUtil } from '../utils/settingsUtil';
+import { Connection, ConnectionInfo } from './connection';
+
+/**
+ * @description keeps track of all of the Connection objects and the info needed to create them (ie: initial ports).
+ */
+const localize = nls.loadMessageBundle();
+export class ConnectionManager extends Disposable {
+	private _initHttpPort;
+	private _initWSPort;
+	private _initHost: string;
+	private _connections: Map<vscode.Uri | undefined, Connection>; // undefined key means no workspace root
+
+	constructor(private readonly _extensionUri: vscode.Uri) {
+		super();
+
+		this._initHttpPort = SettingUtil.GetConfig(this._extensionUri).portNumber;
+		this._initWSPort = this._initHttpPort + 1;
+		this._initHost = SettingUtil.GetConfig(this._extensionUri).hostIP;
+
+		if (!this._validHost(this._initHost)) {
+			this._showIncorrectHostFormatError(this._initHost);
+			this._initHost = DEFAULT_HOST;
+		} else if (
+			vscode.env.remoteName &&
+			vscode.env.remoteName != '' &&
+			this._initHost != DEFAULT_HOST
+		) {
+			vscode.window.showErrorMessage(
+				localize(
+					'hostCannotConnect',
+					'Cannot use the host "{0}" when using a remote connection. Using default {1}.',
+					this._initHost,
+					DEFAULT_HOST
+				)
+			);
+			this._initHost = DEFAULT_HOST;
+		}
+
+		this._connections = new Map<vscode.Uri, Connection>();
+
+		this._register(
+			vscode.workspace.onDidChangeConfiguration((e) => {
+				if (e.affectsConfiguration(SETTINGS_SECTION_ID)) {
+					this._pendingPort = SettingUtil.GetConfig(
+						this._extensionUri
+					).portNumber;
+					this._pendingHost = SettingUtil.GetConfig(this._extensionUri).hostIP;
+				}
+			})
+		);
+	}
+
+	/**
+	 * @description If setting for the initial port is changed, change the port that servers try first
+	 */
+	private set _pendingPort(port: number) {
+		this._initHttpPort = port;
+		this._initWSPort = port + 1;
+	}
+
+	private set _pendingHost(host: string) {
+		if (this._validHost(host)) {
+			this._initHost = host;
+		} else {
+			this._showIncorrectHostFormatError(host);
+			this._initHost = DEFAULT_HOST;
+		}
+	}
+
+	private _validHost(host: string) {
+		return isIPv4(host);
```

Comment:
Can I use ipv6 hostnames? Probably an issue for a different day but it took some work for us to support this in remote-ssh

Reply:
it's not supported. See https://github.com/microsoft/vscode-livepreview/issues/179